test/langtools/tools/javac/util/IteratorsTest.java

```diff
@@ -0,0 +1,172 @@
+        Iterator<?> emptyCompoundIterator = Iterators.createCompoundIterator(List.of(), Function.identity());
+        Assertions.assertThrows(NoSuchElementException.class, emptyCompoundIterator::next);
+        Assertions.assertThrows(NoSuchElementException.class, emptyCompoundIterator::next);
+    }
+
+    // different ways of obtaining an empty iterator are used to make sure
+    // the compound iterator doesn't depend on (checking) the identity of
+    // any one of them
+    @Test
+    public void intermediateEmptyIterators() {
+        List<Iterator<String>> inputs = List.of(
+                Collections.<String>emptyList().iterator(),
+                Collections.emptyListIterator(),
+                Collections.emptyIterator(),
+                List.of("1").iterator(),
+                List.of("2", "3").iterator(),
+                List.<String>of().iterator(),
+                Collections.<String>emptySet().iterator(),
+                List.of("4", "5").iterator(),
+                com.sun.tools.javac.util.List.<String>nil().iterator());
+        Iterator<String> emptyCompoundIterator = Iterators.createCompoundIterator(inputs, Function.identity());
+        var actual = new ArrayList<String>();
+        emptyCompoundIterator.forEachRemaining(actual::add);
+        assertEquals(List.of("1", "2", "3", "4", "5"), actual);
+    }
+
+    @Test
+    public void recursiveEmpty() {
+        Iterable<Iterator<Object>> inner = () -> Iterators.createCompoundIterator(List.of(), i -> Collections.emptyIterator());
+        Iterator<Object> outer = Iterators.createCompoundIterator(inner, Function.identity());
+        assertFalse(outer.hasNext());
+    }
+
+    @Test
+    public void compoundIterator() {
+        TestConverter<TestIterator<String>, String> c = new TestConverter<>(it -> it);
+        TestIterator<String> test1 = new TestIterator<>(List.of("1").iterator());
+        TestIterator<String> test2 = new TestIterator<>(List.of("2").iterator());
+        Iterator<String> compound = Iterators.createCompoundIterator(List.of(test1, test2), c);
+
+        //nothing should be called before the hasNext or next is called:
+        assertAndResetMaxCalls(c, 0);
+        assertAndResetMaxCalls(test1, 0, 0);
+        assertAndResetMaxCalls(test2, 0, 0);
+
+        //when hasNext is called, should invoke the hasNext delegate once:
+        Assertions.assertTrue(compound.hasNext());
+
+        assertAndResetMaxCalls(c, 1);
+        assertAndResetMaxCalls(test1, 1, 0);
+        assertAndResetMaxCalls(test2, 0, 0);
+
+        Assertions.assertTrue(compound.hasNext());
+
+        assertAndResetMaxCalls(c, 0);
+        assertAndResetMaxCalls(test1, 1, 0);
+        assertAndResetMaxCalls(test2, 0, 0);
+
+        //next may invoke hasNext once:
+        Assertions.assertEquals("1", compound.next());
+
+        assertAndResetMaxCalls(c, 0);
+        assertAndResetMaxCalls(test1, 1, 1);
+        assertAndResetMaxCalls(test2, 0, 0);
```

Comment:
Should the test continue, and check that we end up with calls to the second iterator?